ibis/expr/types/joins.py

```diff
@@ -112,47 +112,18 @@ def disambiguate_fields(
     return fields, collisions, equalities
 
 
-def dereference_mapping_left(chain):
-    # construct the list of join table we wish to dereference fields to
-    rels = [chain.first]
-    for link in chain.rest:
-        if link.how not in ("semi", "anti"):
-            rels.append(link.table)
-
-    # create the dereference mapping suitable to disambiguate field references
-    # from earlier in the relation hierarchy to one of the join tables
-    subs = dereference_mapping(rels)
-
-    # also allow to dereference fields of the join chain itself
-    for k, v in chain.values.items():
-        subs[ops.Field(chain, k)] = v
-
-    return subs
-
-
-def dereference_mapping_right(right):
-    # the right table is wrapped in a JoinTable the uniqueness of the underlying
-    # table which requires the predicates to be dereferenced to the wrapped
-    return {v: ops.Field(right, k) for k, v in right.values.items()}
-
-
-def dereference_sides(left, right, deref_left, deref_right):
-    left = left.replace(deref_left, filter=ops.Value)
-    right = right.replace(deref_right, filter=ops.Value)
-    return left, right
-
-
-def dereference_value(pred, deref_left, deref_right):
-    deref_both = {**deref_left, **deref_right}
-    if isinstance(pred, ops.Comparison) and pred.left.relations == pred.right.relations:
-        left, right = dereference_sides(pred.left, pred.right, deref_left, deref_right)
-        return pred.copy(left=left, right=right)
-    else:
-        return pred.replace(deref_both, filter=ops.Value)
+def dereference_value(pred, deref_left, deref_right, deref_both):
+    return deref_both.dereference(pred)
+    # if isinstance(pred, ops.Comparison) and pred.left.relations == pred.right.relations:
```

Comment:
This code bit is responsible for allowing `a.join(a, [a.col == a.col])`

Reply:
Removing it once I have confirmation about that we are disallowing it.

Reply:
This seems very ambiguous to me and _somewhat_ unlikely. +1 for removing it

cc @jcrist @gforsyth for opinions

Reply:
I am +1 on removing it

Reply:
Also +1. Rip it out.